core/txpool/bundlepool/bundlepool.go

```diff
@@ -0,0 +1,430 @@
+
+	// StateAt returns a state database for a given root hash (generally the head).
+	StateAt(root common.Hash) (*state.StateDB, error)
+}
+
+type BundleSimulator interface {
+	SimulateBundle(bundle *types.Bundle) (*big.Int, error)
+}
+
+type BundlePool struct {
+	config Config
+
+	bundles    map[common.Hash]*types.Bundle
+	bundleHeap BundleHeap
+	mu         sync.RWMutex
+
+	slots uint64 // Number of slots currently allocated
+
+	bundleGasPricer *BundleGasPricer
+	simulator       BundleSimulator
+}
+
+func New(config Config) *BundlePool {
+	// Sanitize the input to ensure no vulnerable gas prices are set
+	config = (&config).sanitize()
+
+	pool := &BundlePool{
+		config:          config,
+		bundles:         make(map[common.Hash]*types.Bundle),
+		bundleHeap:      make(BundleHeap, 0),
+		bundleGasPricer: NewBundleGasPricer(config.BundleGasPricerExpireTime),
+	}
+
+	return pool
+}
+
+func (p *BundlePool) SetBundleSimulator(simulator BundleSimulator) {
+	p.simulator = simulator
+}
+
+func (p *BundlePool) Init(gasTip *big.Int, head *types.Header, reserve txpool.AddressReserver) error {
+	return nil
+}
+
+func (p *BundlePool) FilterBundle(bundle *types.Bundle) bool {
+	for _, tx := range bundle.Txs {
+		if !p.filter(tx) {
+			return false
+		}
+	}
+	return true
+}
+
+// AddBundle adds a mev bundle to the pool
+func (p *BundlePool) AddBundle(bundle *types.Bundle) error {
+	if p.simulator == nil {
+		return txpool.ErrSimulatorMissing
+	}
+
+	price, err := p.simulator.SimulateBundle(bundle)
+	if err != nil {
+		return err
+	}
+	minimalGasPrice := p.bundleGasPricer.MinimalBundleGasPrice()
+	if price.Cmp(minimalGasPrice) < 0 && p.slots+numSlots(bundle) > p.config.GlobalSlots {
+		return txpool.ErrBundleGasPriceLow
+	}
+	bundle.Price = price
+	p.bundleGasPricer.Push(price)
+
+	hash := bundle.Hash()
+	if _, ok := p.bundles[hash]; ok {
+		return txpool.ErrBundleAlreadyExist
+	}
+	for p.slots+numSlots(bundle) > p.config.GlobalSlots {
+		p.drop()
+	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.bundles[hash] = bundle
+	heap.Push(&p.bundleHeap, bundle)
+	p.slots += numSlots(bundle)
+
+	bundleGauge.Update(int64(len(p.bundles)))
+	slotsGauge.Update(int64(p.slots))
+	return nil
+}
+
+func (p *BundlePool) GetBundle(hash common.Hash) *types.Bundle {
+	p.mu.RUnlock()
+	defer p.mu.RUnlock()
+
+	return p.bundles[hash]
+}
+
+func (p *BundlePool) PruneBundle(hash common.Hash) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.deleteBundle(hash)
+}
+
+func (p *BundlePool) PendingBundles(blockNumber *big.Int, blockTimestamp uint64) []*types.Bundle {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	ret := make([]*types.Bundle, 0)
+	for hash, bundle := range p.bundles {
+		// Prune outdated bundles
+		if (bundle.MaxTimestamp != 0 && blockTimestamp > bundle.MaxTimestamp) ||
+			blockNumber.Cmp(new(big.Int).SetUint64(bundle.MaxBlockNumber)) > 0 {
+			p.deleteBundle(hash)
+			continue
+		}
+
+		// Roll over future bundles
+		if bundle.MinTimestamp != 0 && blockTimestamp < bundle.MinTimestamp {
```

Comment:
Do we control that `MinTimestamp` can not be too far from now, otherwise it will always occupy the pool.

Reply:
Done